### ISSUE TITLE
RPC failure status should include all useful information about the failure

### ISSUE THREAD

**Issue by markdroth** [grpc/grpc#22883]
See discussion in #21308.

**Comment by markdroth** [comment]
Another interesting case we should consider: If the resolver never returns a result and RPCs fail due to DEADLINE_EXCEEDED, the error message should indicate that the resolver never returned a result.

**Comment by apolcyn** [comment]
> Another interesting case we should consider: If the resolver never returns a result and RPCs fail due to DEADLINE_EXCEEDED, the error message should indicate that the resolver never returned a result.

I agree, I also think this "waiting to acquire resource X" would be helpful debug info for a lot of other things that are currently buried in logs. For example, there are a number of interesting things that an RPC can "stall" on:
* RPC can be stalled because the channel's LB policy hasn't picked a connection for the RPC
  - the connection-pick might be incomplete because:
     a) name resolution hasn't completed
     b) load balancer isn't responding (if relevant to the LB policy in use)
     c) connections are all in "connecting" state.
     ... etc.
* RPC can get stuck waiting for a slot in it's http/2 connection's "max concurrent streams" queue.
* RPC can have an active send message that it hasn't yet written to the wire because it's waiting for space in the outgoing HTTP/2 flow control window
.....etc.

If these things could represented e.g. with some counters, and be accounted for on a per-RPC basis, then they could then be tied back to each RPC's context and be surfaced up to users either in an RPC's error message or perhaps alongside it - IMO that would useful to debug why some RPC's fail and potentially also why some RPC's are slow (e.g. determining that some specific high latency RPC took a long time because it was spending all of it's time waiting for name resolution to finish).

I'm not sure if perhaps the census context could be leveraged for these kinds of things....

**Comment by markdroth** [comment]
This issue is only about things that cause RPCs to fail, so it would not include things like flow control.  Those things are better diagnosed by (e.g.) the application noticing that a send_message op takes too long to complete.

I don't think we need to use census for this, and I don't think we need a new API; I think it's sufficient to just include the relevant information in the status message we return along with the status code (typically UNAVAILABLE for these cases).

But regardless of what API we use to return the information, the important thing is to structure our internals such that all of the information is available at the point at which the status is being generated.

**Comment by markdroth** [comment]
Another example: When a wait_for_ready RPC hits its deadline, the status message should include the error from the last LB pick.

**Comment by markdroth** [comment]
In most LB policies, when we set state to TRANSIENT_FAILURE, we probably want to return just the leaf status (i.e., the status from the last child policy or subchannel that reported a connection failure).  May (or may not) want to do something different for the xds_routing policy, since it may be useful to report which route we attempted to take in addition to the leaf error.

**Comment by connect2satz** [comment]
Hi Mark,
when can we expect the fix for this issue ?

**Comment by markdroth** [comment]
We have no specific ETA.  It's something we're working on slowly over time.

**Comment by connect2satz** [comment]
Hi Mark,
is there any update on the fix ?

**Comment by connect2satz** [comment]
Hi Mark,
     any update on this ?

**Comment by markdroth** [comment]
As I mentioned earlier, we have no specific ETA for this.  It's going to require some internal plumbing changes, and I'm not sure when we'll have a chance to do it.  I'll update this issue as we make progress.

**Comment by connect2satz** [comment]
Hi Mark,
     Is this fixed ?

**Comment by markdroth** [comment]
@ejona86 @dfawley We should write a document of what information is useful or is not useful to include in which circumstances.

A few examples from our meeting today:

For DEADLINE_EXCEEDED:
- how long the RPC was pending before it was failed
- how long the RPC was waiting for name resolution or LB pick
- whether the RPC was still waiting for a connection when it was failed
- whether the RPC was wait_for_ready
- if the RPC had a successful LB pick, the remote address to which it was sent (and the source address from which it was sent)

For UNAVAILABLE:
- If failed due to (e.g.) HTTP status code, include the HTTP status code and maybe some of the body.

**Comment by markdroth** [comment]
Once #30283 is merged, I think we'll be generating useful RPC status messages in all of our LB policies.

The remaining work here is to improve the messages for DEADLINE_EXCEEDED cases and to provide better messages for UNAVAILABLE statuses generated by the transport.

**Comment by markdroth** [comment]
@ananda1066 is going to work on improving the status messages from the transport.
